metadataserver/identity.go

```diff
@@ -41,6 +43,43 @@ func (ms *MetadataServer) generateTaskIdentity() *titus.TaskIdentity {
 	return ti
 }
 
+func (ms *MetadataServer) generateTaskPodIdentity() (*titus.TaskPodIdentity, error) {
+	s := int64(time.Now().Second())     // from 'int'
+	n := int32(time.Now().Nanosecond()) // from 'int'
+	now := &timestamp.Timestamp{Seconds: s, Nanos: n}
+	taskStatus := titus.TaskInfo_RUNNING
+
+	if ms.pod == nil {
+		return nil, fmt.Errorf("No pod object available to marshal as a task identity")
+	}
+	podData, err := proto.Marshal(ms.pod)
```

Comment:
Was this the bit that was supposed to be the k8s wire format rather than the proto encoded?

Reply:
Wait, I thought proto-encoded *was* the wire format?

Reply:
Nope, they've got some whole envelope scheme for their wire format with a magic header: https://kubernetes.io/docs/reference/using-api/api-concepts/#protobuf-encoding

From their docs:

> If you want to store or interact with proto-formatted Kubernetes API objects, we recommend using the "official" serialization stack in k8s.io/apimachinery. Directly serializing these types to proto will not result in data that matches the wire format...

Reply:
Oh wow. I missed that on the first read.

Gosh, it is just the two of us here, we are not building an api for the whole world.
I don't think we have any need for that wrapping.
Can we get away with just one layer of serialization? We don't need to literally match the wire format between kubelet and api server here.

Reply:
Yea, I mean, I guess it goes against that k8s recommendation of not using the messages directly, but I kind of feel like it doesn't really apply here? But in that case this serialization to bytes seems unnecessary; should the `TaskPodIdentity` message just embed the `Pod` field as the k8s `Pod` message rather than bytes?

Reply:
That part no, I don't think so, because it would mean... we would have to bump the software any time there is a protobuf change in the pod spec (which we don't control).

But then again, we kinda do. The pod object we are serializing in the first place comes from our golang k8s client libs. 

I'm not super sure. I just don't have that much experience with protobufs to be sure if this is a good idea or now.
Maybe @sargun might now what is best here?

Certainly putting the pod as bytes in this field I think is "safe". The pod will change, and when it does, we won't have to rebuild any of *our* protos at all. Even you won't either. Your side will have to upgrade to later version of k8s client libs as we upgrade (+/- one version as we go), but that is inevitable if you are sucking in pod objects.

Double wrapping the pod object in another proto? That seems not needed for us here. Heck we are *already* doing that with this task ident proto!

Reply:
Making the field `bytes` or `Pod` yields the same thing on the wire so I suppose it doesn't actually make a difference to how we'll consume it. It would just save you a step serializing to bytes explicitly.

Reply:
So, I strongly recommend not using proto, as the proto serialization / deserialization isn't 100% compatible with the Kubernetes serialization / deserialization (for example, IIRC, you can have defaults in the kubernetes spec). 

I think (and I say this with a sense of depression), the easiest way to do this is to turn TaskPodIdentity into a CRD, and register it, and use the Kubernetes serialization libraries.

Reply:
Waaaaa?

This sounds way overkill. We don't wire-protocol compatibility here, we are not talking to the api server. 
Heck I could just `cat pod.json` and stick that in the `bytes`! All we really need is some way for someone to see what the current pod object is.

Reply:
You definitely can serialize the pod object to json (or literally cat pod.json), and stuff it into a string / bytes. I think that's a good idea.

Reply:
JSON-ified pod stuffed into bytes seems reasonable, as long as @JackOfMostTrades is cool with that.

Reply:
As long as I can parse it back out easily enough in Java I'm fine with that. Hopefully using some k8s library so that I can get it into a POJO. I'm guessing that this would just match the proto JSON format?

Reply:
`POJO` ?

Reply:
[Plain old java object](https://en.wikipedia.org/wiki/Plain_old_Java_object#JavaBeans)

Reply:
This approach sounds good to me.